### PR TITLE
[FEATURE] filter spectra by similarity

### PR DESCRIPTION
**Pull request by timosachsenberg**
An initial version of MS2 spectra filter capability in FileFilter.
Open for suggestions. Tests will be added.

### CODE REVIEW

**Comment by timosachsenberg** [comment]
Added tests

**Comment by timosachsenberg** [comment]
only StopWatch failed...

File: src/topp/FileFilter.cpp [hunk 1]
```diff
@@ -835,23 +851,45 @@ class TOPPFileFilter :
...
+        ConsensusMap consensus_map;
+        ConsensusXMLFile cxml_file;
+        cxml_file.load(consensus_blackorwhitelist, consensus_map);
+        consensus_map.sortByMZ();
+
+        int ret = filterByBlackOrWhiteList(is_blacklist, exp, consensus_map, rt_tol, mz_tol, is_ppm, maps);
+        if (ret != EXECUTION_OK) { return (ExitCodes)ret; }
       }
 
+      // filter spectra if the occur in spectra:blackorwhitelist:file in rt/mz/similarity tolerance window
```
Comment (enetz) [inline src/topp/FileFilter.cpp]: spelling error in this comment

File: src/topp/FileFilter.cpp [hunk 2]
```diff
@@ -1274,6 +1308,82 @@ class TOPPFileFilter :
...
+          writeDebug_("Similarity score: " + String((*comp_function)(exp_spectrum, lib_spectrum)), 10);
+
+          // we have matching spectra
+          list_idx.insert(exp_index); 
+        }
+      }
+    }
+
+    // create new experiment
+    PeakMap exp2 = exp; // copy meta data
```
Comment (hroest) [inline src/topp/FileFilter.cpp]: probably faster:

`PeakMap exp2 = (SpectrumSettings)exp;`

**Comment by hroest** [comment]
:+1: looks good

File: src/topp/FileFilter.cpp (hunk 2 above)
Comment (timosachsenberg) [reply src/topp/FileFilter.cpp]: ah yes. good point

File: src/topp/FileFilter.cpp [hunk 3]
```diff
@@ -1281,8 +1281,8 @@ class TOPPFileFilter :
     }
 
     // create new experiment
-    PeakMap exp2 = (SpectrumSettings)exp; // copy meta data
-    exp2.clear(false); // clear spectra
+    PeakMap exp2;
+    exp.getExperimentalSettings() = (ExperimentalSettings)exp.getExperimentalSettings(); // copy meta data
```
Comment (hroest) [inline src/topp/FileFilter.cpp]: @timosachsenberg -- wait, this should be `exp2` here, not `exp`, right?

Comment (timosachsenberg) [reply src/topp/FileFilter.cpp]: Thanks!

**Pull request merged**